crates/ghost_actor/src/ghost_mutex.rs

```diff
@@ -0,0 +1,197 @@
+            }
+            None => (),
+        }
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> Deref for GhostMutexGuard<'lt, T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &*self.guard.as_ref().expect("exists")
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> DerefMut for GhostMutexGuard<'lt, T> {
+    fn deref_mut(&mut self) -> &mut T {
+        &mut *self.guard.as_mut().expect("exists")
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> AsRef<T> for GhostMutexGuard<'lt, T> {
+    fn as_ref(&self) -> &T {
+        self.deref()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> AsMut<T> for GhostMutexGuard<'lt, T> {
+    fn as_mut(&mut self) -> &mut T {
+        self.deref_mut()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> Borrow<T> for GhostMutexGuard<'lt, T> {
+    fn borrow(&self) -> &T {
+        self.deref()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt> BorrowMut<T> for GhostMutexGuard<'lt, T> {
+    fn borrow_mut(&mut self) -> &mut T {
+        self.deref_mut()
+    }
+}
+
+impl<'lt, T: ?Sized + 'lt + std::fmt::Debug> std::fmt::Debug for GhostMutexGuard<'lt, T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        if f.alternate() {
+            write!(f, "{:#?}", self.deref())
+        } else {
+            write!(f, "{:?}", self.deref())
+        }
+    }
+}
+
+pub struct GhostMutex<T: ?Sized> {
+    mutex: parking_lot::Mutex<T>,
+}
+
+impl<T> GhostMutex<T> {
+    pub fn new(t: T) -> Self {
+        Self {
+            mutex: parking_lot::Mutex::new(t),
+        }
+    }
+
+    pub fn into_inner(self) -> T {
+        self.mutex.into_inner()
+    }
+}
+
+impl<T: ?Sized> GhostMutex<T> {
+    pub fn lock(&self) -> GhostMutexGuard<'_, T> {
+        match self
+            .mutex
+            .try_lock_for(std::time::Duration::from_millis(10))
+        {
+            None => panic!("failed to obtain lock within timeout"),
+            Some(g) => GhostMutexGuard { guard: Some(g) },
+        }
+    }
+
+    pub fn get_mut(&mut self) -> &mut T {
+        self.mutex.get_mut()
+    }
+}
+
+impl<T> From<T> for GhostMutex<T> {
+    fn from(t: T) -> Self {
+        GhostMutex::new(t)
+    }
+}
+
+impl<T: ?Sized + Default> Default for GhostMutex<T> {
+    fn default() -> Self {
+        GhostMutex::new(Default::default())
+    }
+}
+
+impl<T: ?Sized + std::fmt::Debug> std::fmt::Debug for GhostMutex<T> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let mut f = f.debug_struct("GhostMutex");
+        let f = match self.mutex.try_lock() {
+            Some(guard) => f.field("inner", &&*guard),
+            None => f.field("inner", &"<locked>".to_string()),
+        };
+
+        f.finish()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::sync::Arc;
+
+    #[test]
+    fn it_should_debug() {
+        let a = GhostMutex::new(true);
+        assert_eq!("GhostMutex { inner: true }", &format!("{:?}", a));
+    }
+
+    #[test]
+    fn it_should_be_fair_ish() {
```

Comment:
I'm not sure to understand what this test is about :thinking:

Reply:
With the rust system Mutex, if a thread unlocks a mutex, then requests a lock again before the next system switch it will receive the lock, leading to a single thread preference.

Parking lot has a fair unlock option that tries to let other threads get a chance to run.